sdk/storage/azure-storage-blob-nio/src/test/java/com/azure/storage/blob/nio/NioBlobInputStreamTest.groovy

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob.nio
+
+import com.azure.storage.blob.BlobClient
+import spock.lang.Unroll
+
+class NioBlobInputStreamTest extends APISpec {
+
+    File sourceFile
+    BlobClient bc
+    NioBlobInputStream nioStream
+    FileInputStream fileStream
+
+    def setup() {
+        sourceFile = getRandomFile(10 * 1024 * 1024)
+        cc.create()
+        bc = cc.getBlobClient(generateBlobName())
+        bc.uploadFromFile(sourceFile.getPath())
+        nioStream = new NioBlobInputStream(bc.openInputStream())
+        fileStream = new FileInputStream(sourceFile)
+    }
+
+    def cleanup() {
+        sourceFile.delete()
+        cc.delete()
+    }
+
+    def "Read whole file"() {
+        expect:
+        compareInputStreams(nioStream, fileStream, sourceFile.size())
+    }
+
+    def "Read min"() {
+        expect:
+        for (i in 1..100) {
+            assert nioStream.read() == fileStream.read()
+        }
+    }
+
+    @Unroll
+    def "Read buff"() {
+        setup:
+        def nioBytes = new byte[size]
+        def fileBytes = new byte[size]
+
+        when:
+        nioStream.read(nioBytes)
```

Comment:
Does our implementation of BlobInputStream always give back the user as much data as they requested? Cause isnt it possible for an InputStream to return less than expected?

Reply:
You're right that the docs say it could return less, but our read method will always return the amount asked for unless there's an error or the end of the blob is hit. 

```
int chunks = (int) (Math.ceil((double) len / (double) this.chunkSize));
        int numOfBytesRead = 0;
        for (int i = 0; i < chunks; i++) {
            int results = this.readInternal(b, off + numOfBytesRead, len);
            if (results == -1) {
                return -1;
            }
            numOfBytesRead += results;
        }
        return numOfBytesRead;
```